### PR TITLE
Adding 3 missing cors methods

### PR DESCRIPTION
**Pull request by patrickconroy**
Refs #7402

### CODE REVIEW

**Comment by dereuromark** [comment]
I wonder if an $options array argument would be better than 6 arguments after Request...

**Comment by dereuromark** [comment]
Since we have to stay BC, we need to keep the first 3 for 3.x:
```php
    public function cors(
        Request $request,
        $allowedDomains,
        $allowedMethods = [],
        $allowedHeaders = [],
        array $options = []
    ) {
```

And in 4.x we can further reduce it to only 3 arguments in total then.

**Comment by patrickconroy** [comment]
Hah, 1 sec while I resurrect the tests I commented out..

File: src/Network/Response.php [hunk 1]
```diff
@@ -1348,23 +1348,48 @@ public function cookie($options = null)
...
             return;
         }
 
-        $allowedDomains = $this->_normalizeCorsDomains((array)$allowedDomains, $request->is('ssl'));
+        // For BC. This can be removed in 4.x when the $allowedDomains, $allowedMethods, $allowedHeaders params are gone.
+        $options += [
+            'allowedDomains' => (array)$allowedDomains,
+            'allowedMethods' => (array)$allowedMethods,
+            'allowedHeaders' => (array)$allowedHeaders
+        ];
```
Comment (markstory) [inline src/Network/Response.php]: Is it useful to use func_num_args() and accept option in the 2nd argument? Another option could be to add methods for each cors related feature?

File: tests/TestCase/Network/ResponseTest.php [hunk 2]
```diff
@@ -1120,26 +1155,26 @@ public function corsData()
...
+            // [$secureRequest, 'https://www.bar.com', 'www.bar.com', '', '', false, false, false, 'https://www.bar.com', false],
+            // [$secureRequest, 'https://www.bar.com', 'http://www.bar.com', '', '', false, false, false, false, false],
+            // [$secureRequest, 'https://www.bar.com', '*.bar.com', '', '', false, false, false, 'https://www.bar.com', false],
+
+            // [$fooRequest, 'http://www.foo.com', '*', 'GET', '', false, false, false, '*', 'GET'],
+            // [$fooRequest, 'http://www.foo.com', '*.foo.com', 'GET', '', false, false, false, 'http://www.foo.com', 'GET'],
+            // [$fooRequest, 'http://www.foo.com', '*.foo.com', ['GET', 'POST'], '', false, false, false, 'http://www.foo.com', 'GET, POST'],
+            // [$fooRequest, 'http://www.foo.com', '*', '', 'X-CakePHP', false, false, false, '*', false, 'X-CakePHP'],
+            // [$fooRequest, 'http://www.foo.com', '*', '', ['X-CakePHP', 'X-MyApp'], false, false, false, '*', false, 'X-CakePHP, X-MyApp'],
+            // [$fooRequest, 'http://www.foo.com', '*', ['GET', 'OPTIONS'], ['X-CakePHP', 'X-MyApp'], false, false, false, '*', 'GET, OPTIONS', 'X-CakePHP, X-MyApp'],
```
Comment (markstory) [inline tests/TestCase/Network/ResponseTest.php]: Commenting out tests is something I would avoid.

**Comment by lorenzo** [comment]
I think separate methods will be easier to manage

**Comment by patrickconroy** [comment]
Should I write a new class for this then? Six methods in `Response` just for cors would seem wrong to me. If not, should the methods be `Response::corsAllowedDomains()` .. etc?

File: tests/TestCase/Network/ResponseTest.php (hunk 2 above)
Comment (patrickconroy) [inline tests/TestCase/Network/ResponseTest.php]: I was adding a set for the new options (the options for that test are cumbersome), and meant to uncomment those when my new one passed.

**Comment by markstory** [comment]
Does something like:

```php
$response->cors($request)
  ->domains(..)
  ->authentication(true)
  ->headers(['x-csrf-token']);
```
seems reasonable? We could use a small CorsBuilder that takes a response object, and  and sets the headers, or doesn't based on the Origin request header?

**Comment by lorenzo** [comment]
That looks reasonable to me

**Comment by patrickconroy** [comment]
I'll be able to take a stab at this early next week. Should the class be something like `Cake\Network\Response\CORSBuilder` ?

**Comment by dereuromark** [comment]
We usually use name schemes like `CorsBuilder`, though.

**Comment by markstory** [comment]
ping

**Comment by markstory** [comment]
Closing due to inactivity. If this gets updated, we can reopen :smile:

**Pull request closed without merging**